### PR TITLE
Signature

### PR DESCRIPTION
**Pull request by termim**
wait_for and run_inreactor decorators do not preserve signature which makes
automatic documentation (ex. Sphinx autodoc) more difficult. These changes use decorator module [1] to fix that.

[1] https://pypi.python.org/pypi/decorator

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 Twisted>=11.1.0
 sphinx
+decorator
```
Comment (itamarst) [inline requirements-dev.txt]: This isn't a dev requirement, it's actually needed to install the package, so it should go in `setup.py`.

File: crochet/tests/test_api.py [hunk 2]
```diff
@@ -757,6 +772,18 @@ def passthrough(argument):
...
+    def test_signature(self):
+        """
+        The function decorated with the wait decorator has the same signature
+        as the original function.
+        """
+        decorator = self.decorator()
+        def some_name(arg1, arg2, karg1=2, *args, **kw):
+            pass
+        decorated = decorator(some_name)
+        self.assertEqual(str(inspect.getargspec(some_name)),
```
Comment (itamarst) [inline crochet/tests/test_api.py]: No need to do `str`

File: crochet/tests/test_api.py [hunk 3]
```diff
@@ -609,6 +610,20 @@ class RunInReactorTests(TestCase):
...
+        """
+        The function decorated with the run_in_reactor decorator has the same
+        signature as the original function.
+        """
+        c = EventLoop(lambda: FakeReactor(), lambda f, g: None)
+
+        def some_name(arg1, arg2, karg1=2, *args, **kw):
+            pass
+        decorated = c.run_in_reactor(some_name)
+        self.assertEqual(str(inspect.getargspec(some_name)),
```
Comment (itamarst) [inline crochet/tests/test_api.py]: No need to do `str`

**Comment by itamarst** [comment]
Thanks, that's a good idea.

1. Add news entry in documentation, don't forget to thank yourself.
2. Fix comments above; Travis-CI build should then pass other than known issue with PyPy (#73).

**Comment by itamarst** [comment]
Or, if you feel like you've done enough I'd be happy to take over. Let me know.

**Comment by termim** [comment]
I fixed your comments above. Would you please check what else is missing.

**Comment by itamarst** [comment]
Looks good! Merge forward so Travis-CI and GitHub are happy and I'll take one final look and then likely merge to master as is.

**Comment by termim** [comment]
Done.

**Comment by itamarst** [comment]
This looks pretty good. The only thing that worries me is that, now that I check the docs, "decorators generated by `decorator` can only be used on user-defined Python functions or methods, not on generic callable objects, nor on built-in functions, due to limitations of the inspect module in the standard library."

This means this is not backwards compatible, and somewhat limiting in what can be decorated. The `wrapt` library appears to provide the same benefits of `decorator` re preserving signature, while having fallback code to support generic callables. It also has a few other improvements over `decorator`. Could you use it instead? Thanks!

**Comment by itamarst** [comment]
Another worry is that using `wrapt` will break putting `@wait_for` below `@classmethod`; `wrapt` appears to require putting it above `@classmethod`, but right now only putting it below works (see #55). Some investigation suggests that for our purposes it will actually work fine, but I suggest writing a couple tests for both variants of `@classmethod` interaction just in case (above and below). If both work this can also close #55.

**Comment by itamarst** [comment]
Going to see if I can do this with `wrapt`.

**Comment by itamarst** [comment]
OK, superseded by my branch.

**Pull request closed without merging**